### PR TITLE
Psp 6322 - lease and license refactor

### PR DESCRIPTION
**Pull request by devinleighsmith**

### CODE REVIEW

File: source/backend/api/Areas/Leases/Controllers/LeaseController.cs [hunk 1]
```diff
@@ -94,11 +78,11 @@ public IActionResult AddLease(Api.Models.Concepts.LeaseModel leaseModel, [FromQu
...
-        /// <summary>
-        /// Update the specified lease, and attached properties.
-        /// </summary>
-        /// <returns></returns>
-        [HttpPut("{id:long}")]
+    /// <summary>
+    /// Update the specified lease, and attached properties.
+    /// </summary>
+    /// <returns></returns>
+    [HttpPut("{id:long}")]
```
Comment (FuriousLlama) [inline source/backend/api/Areas/Leases/Controllers/LeaseController.cs:85]: nit: indentation seems off

File: source/backend/api/Areas/Leases/Controllers/LeaseController.cs [hunk 2]
```diff
@@ -51,22 +51,6 @@ public LeaseController(ILeaseService leaseService, IMapper mapper)
...
-        /// <summary>
-        /// Get the lease for the specified primary key 'id'.
-        /// </summary>
-        /// <returns></returns>
-        [HttpGet("concept/{id:long}")]
-        [HasPermission(Permissions.LeaseView)]
-        [Produces("application/json")]
         [ProducesResponseType(typeof(IEnumerable<Api.Models.Concepts.LeaseModel>), 200)]
         [SwaggerOperation(Tags = new[] { "lease" })]
         public IActionResult GetLeaseConcept(int id)
```
Comment (FuriousLlama) [inline source/backend/api/Areas/Leases/Controllers/LeaseController.cs:56]: nit: now that there is only the lease concept, maybe we can rename this method

File: source/backend/api/Areas/Leases/Controllers/LeaseTenantController.cs [hunk 3]
```diff
@@ -21,42 +26,77 @@ namespace Pims.Api.Areas.Lease.Controllers
...
+            var updatedLease = _leaseService.GetTenantsByLeaseId(leaseId);
+
+            return new JsonResult(_mapper.Map<IEnumerable<LeaseTenantModel>>(updatedLease));
+        }
+
         /// <summary>
         /// Update the specified tenants on the passed lease.
         /// </summary>
         /// <returns></returns>
         [HttpPut("{leaseId:long}/tenants")]
```
Comment (FuriousLlama) [inline source/backend/api/Areas/Leases/Controllers/LeaseTenantController.cs:81]: nit: a tad more restful would be to pass the tenant id as well

Comment (FuriousLlama) [reply source/backend/api/Areas/Leases/Controllers/LeaseTenantController.cs:81]: oh, nvm i thought it was for a single tenant

File: source/backend/api/Areas/Leases/Controllers/LeaseTermController.cs [hunk 4]
```diff
@@ -33,31 +39,65 @@ public class LeaseTermController : ControllerBase
...
         }
         #endregion
 
         #region Endpoints
 
+        /// <summary>
+        /// Update the specified term on the passed lease.
+        /// </summary>
+        /// <returns></returns>
+        [HttpGet("{leaseId:long}/term")]
```
Comment (FuriousLlama) [inline source/backend/api/Areas/Leases/Controllers/LeaseTermController.cs:58]: nit: most of our paths are plural

File: source/backend/api/Areas/Leases/Controllers/SecurityDepositController.cs [hunk 5]
```diff
@@ -77,63 +116,85 @@ public IActionResult AddDeposit(long leaseId, [FromBody] ParentConcurrencyGuardM
...
         /// <summary>
         /// Delete the specified deposit on the passed lease.
         /// </summary>
-        /// <returns></returns>
-        [HttpDelete("{leaseId:long}/deposits")]
+        [HttpDelete("{leaseId:long}/deposits/{depositId:long}")]
         [HasPermission(Permissions.LeaseEdit)]
         [Produces("application/json")]
-        [ProducesResponseType(typeof(IEnumerable<LeaseModel>), 200)]
+        [ProducesResponseType(typeof(void), 200)]
```
Comment (FuriousLlama) [inline source/backend/api/Areas/Leases/Controllers/SecurityDepositController.cs:154]: nit: the other delete methods return a boolean.

File: source/backend/dal/Repositories/LeaseRepository.cs [hunk 6]
```diff
@@ -76,23 +76,7 @@ public PimsLease Get(long id)
 
             PimsLease lease = this.Context.PimsLeases.AsSplitQuery()
```
Comment (FuriousLlama) [inline source/backend/dal/Repositories/LeaseRepository.cs:77]: thank you for cleaning this one up!

File: source/frontend/src/features/leases/add/AddLeaseContainer.test.tsx [hunk 7]
```diff
@@ -157,75 +156,9 @@ describe('AddLeaseContainer component', () => {
       userEvent.click(await screen.findByText('Acknowledge & Continue'));
     });
 
-    expect(mockAxios.history.post[0].data).toEqual(JSON.stringify(leaseData));
+    expect(mockAxios.history.post[0].data).toEqual(leaseData);
   });
 });
 
-const leaseData: Api_Lease = {
```
Comment (FuriousLlama) [inline source/frontend/src/features/leases/add/AddLeaseContainer.test.tsx:164]: nit: imo the typed version is more maintainable

File: source/frontend/src/features/leases/add/AddLeaseContainer.tsx [hunk 8]
```diff
@@ -52,9 +52,10 @@ export const AddLeaseContainer: React.FunctionComponent<
     formikHelpers: FormikHelpers<LeaseFormModel>,
     userOverrideCodes: UserOverrideCode[],
   ) => {
-    const leaseApi = leaseFormModel.toApi();
+    const leaseApi = LeaseFormModel.toApi(leaseFormModel);
```
Comment (FuriousLlama) [inline source/frontend/src/features/leases/add/AddLeaseContainer.tsx:55]: q: given that the leaseFormModel has all the information for the transformation to the API version calling it as static is not necessary. What advantage do you see in it being static?

File: source/frontend/src/features/leases/detail/LeasePages/deposits/DepositsContainer.tsx [hunk 9]
```diff
@@ -165,106 +170,99 @@ export const DepositsContainer: React.FunctionComponent<
...
-      let request: IParentConcurrencyGuard<Api_SecurityDepositReturn> = {
-        parentId: lease.id,
-        parentRowVersion: lease.rowVersion,
-        payload: returnDepositForm.toInterfaceModel(),
-      };
-      const updatedLease = await updateLeaseDepositReturn(request);
-      if (!!updatedLease?.id) {
-        setLease(updatedLease);
+    if (lease && lease.id) {
+      let request: Api_SecurityDepositReturn = returnDepositForm.toInterfaceModel();
```
Comment (FuriousLlama) [inline source/frontend/src/features/leases/detail/LeasePages/deposits/DepositsContainer.tsx:174]: nit: toInterfaceModel -> toApi

**Review by FuriousLlama — APPROVED** [review]

File: source/frontend/src/features/leases/add/AddLeaseContainer.tsx (hunk 8 above)
Comment (devinleighsmith) [reply source/frontend/src/features/leases/add/AddLeaseContainer.tsx:55]: There are some problems with non-static functions in classes that we were previously working around with any or as,
primarily it prevents some usages of ellipsis (...) which can be an issue for default* objects as well as test methods.

using static methods allows typescript to treat these models as if they were simple objects which makes transformations easier.

**Pull request merged**